rfc/indexing-design.md

```diff
@@ -0,0 +1,228 @@
+
+Location commitments that describe where content is stored can change over time as w3s moves data to a more favorable location, re-aggregates data storage, or replicates stored data. Location commitments are updated without having to index all multihashes.
+
+A query API will allow a user to specify indexed content to lookup and how to filter results. The API retrieves previously cached responses or composes new IPNI queries to get new responses to filter and cache.
+
+## Overview
+
+A user stores content as opaque blobs containing series of blocks addressed by multihashes. These blocks are collected into one or more CAR file (shards) and the CAR file is stored at a location designated for a particular user's data.
+
+The user creates a blob-index ([sharded-dag-index](https://github.com/w3s-project/specs/blob/main/w3-index.md#sharded-dag-index-example)) that contains the multihash of each shard of blob data, and an array of tuples of (multihash, offset, length) for all the blocks in each shard. The user may choose which blocks to include in the blob-index if they do not want all blocks to be indexed. For example, they may want only the DAG root to be indexed.
+
+The blob-index is stored in the user's space along with the CAR file(s) containing the blob data. This is done in the [w3 add-blob operation](https://github.com/w3s-project/specs/blob/main/w3-blob.md#add-blob). The w3s service creates [location commitments](https://github.com/w3s-project/specs/blob/main/w3-blob.md#location-commitment) for the blob shards and for the blob-index. 
+
+After the add-blob process completes, and the user receives an [accept-blob receipt](https://github.com/w3s-project/specs/blob/main/w3-blob.md#accept-blob-receipt), the user may then choose to make their data publicly queryable by publishing it on [IPNI](https://github.com/ipni/specs/blob/main/IPNI.md). The user optionally invokes an [index-add](https://github.com/w3s-project/specs/blob/main/w3-index.md#index-add) capability to publish the blob-index multihashes to IPNI so that they can be used to look up location commitments and retrieve blob-index information and blob data. See [W3 Index](https://github.com/w3s-project/specs/blob/main/w3-index.md#w3-index) for more.
+
+After publishing to IPNI, any user looking for content can query IPNI for the blob-index of that content. The user read the blob-index to determine what data shard needs to be retrieved, and then asks IPNI for the location of that shard.
+
+Finally the user uses the shard location, with range information from the blob index, to retrieve a specific piece of data from w3s.
+
+## Index Query
+
+A w3s client can query indexed information by specific attributes, and can filter retrieved query results.
+
+The queryable attributes are as follows.
+- Site: the location(s) where all blob data is available
+- Shard: the location(s) where a specific shard of data is available
```

Comment:
`Site`: The idea is that the query can specify "where site is in [X]" so that query response will only contain results that have something stored at a site in [X].

`Shard`: The idea is that a query can specify to retrieve results for requested shards, i.e. "where shard in [S]"

I will try to clarify.